data/provider/src/main/java/tech/pegasys/teku/api/staterootselector/StateRootSelectorFactory.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Consensys Software Inc., 2023
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.api.staterootselector;
+
+import static tech.pegasys.teku.spec.config.SpecConfig.GENESIS_SLOT;
+
+import java.util.Optional;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.api.AbstractSelectorFactory;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.spec.datastructures.blocks.BeaconBlockHeader;
+import tech.pegasys.teku.spec.datastructures.metadata.ObjectAndMetaData;
+import tech.pegasys.teku.spec.datastructures.state.beaconstate.BeaconState;
+import tech.pegasys.teku.storage.client.ChainHead;
+import tech.pegasys.teku.storage.client.CombinedChainDataClient;
+
+public class StateRootSelectorFactory extends AbstractSelectorFactory<StateRootSelector> {
+  private final Spec spec;
+
+  public StateRootSelectorFactory(final Spec spec, final CombinedChainDataClient client) {
+    super(client);
+    this.spec = spec;
+  }
+
+  @Override
+  public StateRootSelector stateRootSelector(final Bytes32 stateRoot) {
+    return () -> client.getStateByBlockRoot(stateRoot).thenApply(this::fromState);
```

Comment:
im confused, is the input to this function a state root or a block root? we've got it getting passed to byblockRoot, which kind of suggests the param is a blockRoot...

Reply:
I'm also not sure we're solving the actual problem in the ticket here, which suggested we could avoid loading the state ,but this still loads state to (compute) get the state root...

Reply:
My bad for `getStateByBlockRoot`. Replaced it by `getStateByStateRoot`

Reply:
Actually as mentioned in the ticket we don't have a slot to state root mapping for the finalized states. I added an optimisation for the `slotSelector` to try to retrieve the state root from the proto array before hitting the database. That's the only optimisation possible for state root requests as far as I can see